File: advent-of-code-2018/src/main/java/eu/janvdb/aoc2018/day13/Direction.java
package eu.janvdb.aoc2018.day13;

enum Direction {
	LEFT, UP, RIGHT, DOWN;

	static Direction fromChar(char ch) {
		return switch (ch) {
			case '<' -> LEFT;
			case '>' -> RIGHT;
			case '^' -> UP;
			case 'v' -> DOWN;
			default -> throw new IllegalArgumentException(String.valueOf(ch));
		};
	}

	Direction turn(Turn turn) {
		return switch (turn) {
			case LEFT -> rotateLeft();
			case RIGHT -> rotateRight();
			case STRAIGHT -> this;
		};
	}

	private Direction rotateLeft() {
		Direction[] values = Direction.values();
		int index = (this.ordinal() + values.length - 1) % values.length;
		return values[index];
	}

	private Direction rotateRight() {
		Direction[] values = Direction.values();
		int index = (this.ordinal() + 1) % values.length;
		return values[index];
	}

	Direction handleCornerTopLeftOrDownRight() { /* / */
		return switch (this) {
			case UP -> RIGHT;
			case RIGHT -> UP;
			case DOWN -> LEFT;
			case LEFT -> DOWN;
		};
	}

	Direction handleCornerTopRightOrDownLeft() {  /* \ */
		return switch (this) {
			case UP -> LEFT;
			case LEFT -> UP;
			case DOWN -> RIGHT;
			case RIGHT -> DOWN;
		};
	}
}

File: kotlin-util/src/main/kotlin/eu/janvdb/aocutil/kotlin/Direction.kt
package eu.janvdb.aocutil.kotlin

enum class Direction {
	N, S, E, W;

	fun rotateLeft(amount: Int): Direction {
		var actualAmount = amount % 360
		if (actualAmount < 0) actualAmount += 360
		return when (amount) {
			0 -> this
			90 -> when (this) {
				N -> W
				W -> S
				S -> E
				E -> N
			}
			180 -> when (this) {
				N -> S
				W -> E
				S -> N
				E -> W
			}
			270 -> when (this) {
				N -> E
				E -> S
				S -> W
				W -> N
			}
			else -> throw IllegalArgumentException(amount.toString())
		}
	}

	fun rotateRight(amount: Int): Direction {
		return rotateLeft(360 - amount)
	}
}